_posts/2020-06-26-Making-IrishGen-SPARQL-Part-I.md

```diff
@@ -0,0 +1,549 @@
+
+The results of this raise an interesting point.  The reader will
+notice that Cathair Már appears three times in the results.  This
+means that there are _three_ different declared number of children for
+Cathair Már and they are not counted the same as `max` should return
+only the largest for a single individual.  To investigate this is
+outside the scope of this post but a few hypothesises suggest
+themselves.  First, that Cathair Már has differing numbers and the
+genealogists were themselves confused while taking from their own
+sources or constructing genealogies with competing interests.  Second,
+these are all the same person but they have not been linked together
+by `owl:sameAs` by the curators which would then match what is
+expected.  Third, there is a discrepancy between LL and Rawl B502 that
+is now apparent by looking at this query.  Whatever the actual
+underlying cause, the query demonstrates something that would need to
+be done painstakingly by hand with an even larger margin for error
+than using a Triplestore to do the calculation for the user.  Large
+aggregate questions about medieval Ireland can now be asked and
+answered with a greater degree of confidence.
+
+There is, of course, another method of counting by taking advantage of
+the reasoner:
+
+```sparql
+prefix irishRel: <http://example.com/earlyIrishRelationship.ttl#>
+prefix rel: <http://purl.org/vocab/relationship/>
+
+select ?x (count(distinct ?y) as ?numChildren)
+from <tag:stardog:api:context:all>
+where {
+    ?x rel:parentOf ?y
+}
+group by ?x
+order by desc(?numChildren)
+limit 5
+```
+
+As a reader familiar with the IrishGen dataset will know
+`rel:parentOf` is not encoded very often within the dataset.  The
+reasoner knows that `rel:parentOf` is the inverse of `rel:childOf` and
+thus can logically infer the number of children by applying this logic
+to the dataset.  Thus, queries can be constructed that are not encoded
+within the dataset.  This does come at a cost that shifts the burden
+of calculating this from human beings encoding it in the dataset to
+the computer but this is a much less labour intensive way of
+determining these things.  In this case, the query is runs for ~1022
+milliseconds (about 1 second).  The outcome of this query for the top
+five people is:
+
+| ?x                                                                              | ?numChildren |
+| http://example.com/Laud_Misc_610/CGH/do_minigud_senchais_fer_muman.trig#Óengus  | 34           |
+| http://example.com/Laud_Misc_610/CGH/senchus_dáil_fíatach.trig#AilellaÁuluim    | 30           |
+| http://example.com/LL/laigsi.trig#CathairMár                                    | 29           |
+| http://example.com/Rawl_B502/úi_meic_h_eirc.trig#ConallClóen                    | 25           |
+| http://example.com/Rawl_B502/do_forslointib_ulad_iar_coitchiund_in_so.trig#Buan | 25           |
```

Comment:
I checked these numbers by expanding the nodes on the visual graph and counting their children. The numbers here are all slightly higher than the numbers returned by that method. Perhaps double-check the query? Or remind the reader that our data is not yet properly proofed, which may return distorted data.